benchmarks/discovery.md

```diff
@@ -0,0 +1,148 @@
+protocol](../ssdp.md).
+2. Write our own SSDP implementation from scratch, possibly based on the
+   [Chromium](https://cs.chromium.org/chromium/src/chrome/browser/media/router/discovery/dial/dial_service.h)
+   implementation.
+
+As part of these implementations, we will instrument them to log data related to
+the Observables above, including (but not limited to):
+* Packets and bytes sent and received each second
+* Timestamp a device is discovered
+* Timestamp a device is detected to have left the network
+* Number of devices discovered
+
+A Python script will be used to start the sender and responder process with
+command line flags to adjust behavior for a given test condition. Data points of
+interest will be logged by the binaries to stdout, which will be injested and
+accumulated by the Python script.  At the end of the test, the script will write
+a JSON file to be copied to a Web server, so that a HTML frontend can be used to
+render nice charts of the results.  All of the scripts and source code for the
+benchmarks will be stored on GitHub.
+
+# Experimental Variables
+
+The following variables will be adjusted and the Observables collected for each
+condition.
+
+To keep the number of conditions within reason, one set of parameters will be
+adjusted and the others kept fixed.  In case the data shows interesting trends,
+we can run more conditions as desired as the entire process will be automated
+through scripting.
+
+## Protocols in use
+
+* SSDP alone
+* mDNS alone
+* SSDP and mDNS together
+
+## Device Population
+
+* Number of senders (devices initiating queries): 1, 2, 4, 8
+* Number of responders (devices responding to queries): 1, 2, 4, 8
+
+## Physical Positioning
+
+Most benchmarks will take place with the devices co-located (same room, same
+distance from router).
+
+However, to simulate signal fading effects, we will also collect data when the
+devices will be positioned in adjacent offices such that the senders are in one
+room separated by a wall from the router, and the responders are in a different
+room separated by a wall from the router.  This is to simulate a typical
+household where controller devices, receiver devices, and routers are spread
+among different rooms.
+
+**NOTE:** If there is interest in evaluating these protocols in particularly
+adverse WiFi conditions, we can investigate using a routing proxy to simulate
+additional congestion and packet loss, or simply move devices further away.
+
+## SSDP specific
+
+* Interval between `M-SEARCH` requests.
+
+## mDNS specific
+
+* TTL assigned to mDNS answers.
+
+# Data Collection And Analysis
+
+For numerical values, data will be collected for a given experimental condition
+N = 30 times and the 50%, 75%, and 95% values computed, as well as time series.
+
+One experimental condition (the control) will have no senders or responders to
+collect baseline data.
+
+# Notes
+
+Power consumption measurement only makes sense as a relative measurement, and
```

Comment:
We can track battery level, number of discovery requests, number of found devices for both protocols mDNS and SSDP with the same number of senders and responders in the network. The experiment starts with fully charged smartphone and ends until the battery level reaches a pre-defined level e.g. 50%.

Reply:
That would be one approach, although I worry that the amount of battery consumed is so small relative to the other power sinks on a mobile device that it will get lost in the noise.  There are a lot of mobile-specific implementation techniques that can affect power like coalescing background tasks and/or radio wakeups for passive discovery mechanisms like these.

Added a note to look at battery levels as a first approximation; but this really requires some more research and I will file a task to follow up.